### ISSUE TITLE
Rust, gpu and google colab

### ISSUE THREAD

**Issue by maurokenny** [evcxr/evcxr#147]
Hi guys,

It is possible to use rust in google colab? In addition, is it possible to use also a gpu there with rust?

Thanks

**Comment by davidlattimore** [comment]
Sorry about the slow reply. It is possible to use with colab. It's a super streamlined process though. You need a to build the evcxr_jupyter binary, then upload it to Google Drive and then reference the Google Drive file. See [this comment](https://github.com/google/evcxr/issues/112#issuecomment-609366627). I'd been meaning to try to figure out if the process could be streamlined somewhat. e.g. do automatic builds pushed to Google Drive.

As far as using the GPU, not sure sorry. I haven't done GPU programming from Rust or otherwise.

**Comment by melMass** [comment]
> It is possible to use rust in google colab? In addition, is it possible to use also a gpu there with rust?

Yes and yes. Check this [gist](https://gist.github.com/korakot/ae95315ea6a3a3b33ee26203998a59a3), in the comments korakot provides a starting notebook (don't create one from scratch see comments for details)

**Comment by gabelluardo** [comment]
Seems that the gist doesn't work anymore

**Comment by davidlattimore** [comment]
From the limited reading I just did online, it seems that kernels other than Python 3 aren't supported by Colab. I found a link claiming to create a Swift Colab, but it too didn't seem to work. Perhaps it was possible before with some hacks but those hacks have stopped working with recent versions of Colab. I attempted to repeat those hacks just now, but if I try to execute Rust code, I get `Failed to execute cell. Could not send execute message to runtime: Error: await connected: disconnected`. Probably not much more that can be done at this point.

**Comment by Panakotta00** [comment]
I've got it working again (kinda hacky tho), can be found in a comment of [this gist](https://gist.github.com/korakot/ae95315ea6a3a3b33ee26203998a59a3).

**Comment by eatPizza311** [comment]
@wiseaidev Thank you for your amazing contribution.
The script works perfectly on CPU runtime, but when I tried on runtime with GPU it's not working.
Here is the error message:
![圖片](https://github.com/evcxr/evcxr/assets/43287234/1b4fb3c5-6751-496d-91bf-e5b5bca7f990)
After some digging, I found the symbolic link was correctly created, but the command is only working with the absolute path.
The configuration works perfectly if I prefixed the `nix-dev` and `evcxr_jupyter` with `/root/.nix-profile/bin` until line 26:
`bash -c "killall jupyter-notebook ; sleep 3 ; source ~/.cargo/env && jupyter notebook --ip=172.28.0.12 --port=9000" </dev/null>/dev/null &`
I've tried to change the runtime manually, and the backend seems not able to connect.
Do you have time to take a look at this?

**Comment by xavierforge** [comment]
Hi @davidlattimore, following up on my comment from March 2024. I ended up building this.

https://github.com/xavierforge/colab-rust

It's the auto-build idea you floated in 2020, but with GitHub Actions and Releases instead of Google Drive, so the URLs are stable, there's version history, it rebuilds weekly and needs no auth. On Colab, setup goes from about 11 minutes with `cargo install` to about 17 seconds.

The binary is built on an `ubuntu-22.04` runner, so it runs on both Colab's old 22.04 image and the current 24.04 one, and it falls back to `cargo install --locked` if the download fails. It stays on the Python runtime. A `%%rust` cell magic runs `evcxr_jupyter` as a subprocess through `jupyter_client`, so there's no kernel switch, it works on GPU runtimes, and Python and Rust cells can mix in one notebook. The stop button sends a real `interrupt_request`, which has been clean since 0.22.0 (thanks for #440). Limitations are that output is buffered per cell and highlighting is Python's.

For the record, the gist's setup cell breaks on GPU runtimes because that image already has `/opt/bin` for NVIDIA tools, so `ln -s /root/.nix-profile/bin /opt/bin` creates `/opt/bin/bin` instead and `nix-env` is never on PATH. That's also why only absolute paths worked in my 2024 comment. Even with that patched, the kernel switch is fragile, which is what pushed me to the subprocess design.

If you think the original question is answered now, I'd be glad to see this closed. Either way, thanks for evcxr and for merging #506.

**Comment by davidlattimore** [comment]
Thanks for looking into this!
